### ISSUE TITLE
regression in `1.1.350` with `UnionType`

### ISSUE THREAD

**Issue by samuelcolvin** [microsoft/pyright#7228]
The pyright behaves differently with the following code between 1.1.343 and 1.1.350, the previous behavior was correct.

```py
from typing import Generic, TypeVar

T = TypeVar('T')


class Foobar(Generic[T]):
    def __init__(self, value_type: type[T]) -> None:
        self.value_type = value_type

    def make_value(self) -> T:
        raise NotImplementedError


f = Foobar(int)

reveal_type(f.make_value())
# above is correct in both 1.1.343 and 1.1.350
# > information: Type of "f.make_value()" is "int"

f = Foobar(int | str)

reveal_type(f.make_value())
# correct in 1.1.343
# > information: Type of "f.make_value()" is "int | str"
# but wrong in 1.1.350
# > information: Type of "f.make_value()" is "Unknown"
```

## Full output with `v1.1.343`

```
WARNING: there is a new pyright version available (v1.1.343 -> v1.1.350).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

/Users/samuel/code/pydantic-data-platform/scratch.py
  /Users/samuel/code/pydantic-data-platform/scratch.py:16:13 - information: Type of "f.make_value()" is "int"
  /Users/samuel/code/pydantic-data-platform/scratch.py:23:13 - information: Type of "f.make_value()" is "int | str"
0 errors, 0 warnings, 2 informations 
```

## Full output of `v1.1.350`

```
/Users/samuel/code/pydantic-data-platform/scratch.py
  /Users/samuel/code/pydantic-data-platform/scratch.py:16:13 - information: Type of "f.make_value()" is "int"
  /Users/samuel/code/pydantic-data-platform/scratch.py:21:1 - error: Type of "f" is partially unknown
    Type of "f" is "Foobar[Unknown]" (reportUnknownVariableType)
  /Users/samuel/code/pydantic-data-platform/scratch.py:21:12 - error: Argument of type "type[int] | type[str]" cannot be assigned to parameter "value_type" of type "type[T@Foobar]" in function "__init__"
    Type "UnionType" cannot be assigned to type "type[T@Foobar]" (reportArgumentType)
  /Users/samuel/code/pydantic-data-platform/scratch.py:23:13 - information: Type of "f.make_value()" is "Unknown"
2 errors, 0 warnings, 2 informations 
code/pydantic-data-platform 1 ➤ poetry run pyright --version 
pyright 1.1.350
```

**Comment by Viicos** [comment]
You can find more info in this discussion: https://github.com/microsoft/pyright/discussions/7091
which links to the following [discussion](https://discuss.python.org/t/43898/)

**Comment by erictraut** [comment]
Pyright's current (newer) behavior is correct. The old behavior did not conform to the typing spec. I've been working to increase [pyright's conformance](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html). Pyright's newer behavior is consistent with mypy, which was already conforming with the spec in this regard.

The problem here is that the expression `int | str` has a different meaning when used in a type expression (an annotation) and when used in a value expression. When it's used in a type expression, it evaluates to "a type that represents a union between `int` and `str`". When used in a value expression, it evaluates to a runtime `UnionType` object. The latter is not compatible with `type[T]` according to the typing spec. 

There have been some discussions over the past few years about adding a `TypeForm[T]` to the typing spec that is compatible with any valid type form, but the discussions have stalled, and no one has stepped up to write a formal spec for the concept.

**Comment by EdAyers** [comment]
Minimal example:

```python
from typing import Optional, TypeVar


T = TypeVar("T")


def foo(x: type[T], t: T) -> T:
    return t


foo(Optional[int], None) 
# Argument of type "type[int] | type[None]" cannot be assigned to parameter "x" of type "type[T@foo]" in function "foo"
#   Type "type[UnionType]" cannot be assigned to type "type[None]"
```

I understand that you had to remove this feature, but it makes me sad.

**Comment by JonathanPlasse** [comment]
Would this be enough for your use case?
```python
from typing import TypeVar, overload, reveal_type

T = TypeVar("T")


@overload
def foo(x: type[T], t: None) -> None:
    ...


@overload
def foo(x: type[T], t: T) -> T:
    ...


def foo(x: type[T], t: T | None) -> T | None:
    return t


reveal_type(foo(int, None))  # Type of "foo(int, None)" is "None"
reveal_type(foo(int, 1))  # Type of "foo(int, 1)" is "int"
```

**Comment by EdAyers** [comment]
thanks for the suggestion but sadly not I would need the full `TypeForm` thing.
